### DISCUSSION TITLE
How to check if a user on my app has Patreon membership

### DISCUSSION THREAD

**Discussion by OKNoah** [EddieHubCommunity/support#86] [Q&A]
Has anyone used Patreon as basically subscription payment handler for a website/app?

- Is possible to link users of an app to paying patrons? (in order to give perks)
- With their API. I basically want to use it like a payment processor
- you sign up on my app, decide to become paying user, click link to become patron, link account, and then given extra features within my app so long as you're a patron

- they can become a member on patreon, not through my app
- i just wanna link them there, but wanna know the process (if any) of confirming they're a patron

i remember spending some time on it before and something didn't seem right about it. spoke to someone else who said oh yeah tthey basically don't care abou hte API, etc

**Comment by eddiejaoude** [comment]
Patreon seems to have integrations to many apps (~750) [here](https://zapier.com/apps/patreon/integrations)

But if you want to do a custom integration, you will have to use their API. They seem to have a few SDKs depending on what language you are using. 

I would double check that the SDK does what you want first before integration, as you might have to go straight to the API - that has caught me out before on another API/SDK situation

I have not used Patreon, but from their basic docs, I think the flow would need to be:

1. User logs into your app via Patreon [OAuth](https://docs.patreon.com/#oauth)
2. Now your app knows who the user is on Patreon, you can get their user information and any membership information
3. Then you can give them perks on your app

## User

```typescript
{
  "type": "user"
  "id": <string>
  "attributes": {
    "first_name": <string>
    "last_name": <string>
    "full_name": <string>
    "vanity": <string>
    "email": <string>
    "about": <string>
    "facebook_id": <string>
    "image_url": <string>
    "thumb_url": <string>
    "youtube": <string>
    "twitter": <string>
    "facebook": <string>
    "created": <date>
    "url": <string>
    // optional properties
    "like_count": <int>
    "comment_count": <int>
  }
  "relationships": {
    "campaign": ...<campaign>...
  }
}
```

## Campaign 

```typescript
{
  "type": "campaign"
  "id": <string>
  "attributes": {
    "summary": <string>
    "creation_name": <string>
    "pay_per_name": <string>
    "one_liner": <string>
    "main_video_embed": <string>
    "main_video_url": <string>
    "image_small_url": <string>
    "image_url": <string>
    "thanks_video_url": <string>
    "thanks_embed": <string>
    "thanks_msg": <string>
    "is_monthly": <bool>
    "is_nsfw": <bool>
    "created_at": <date>
    "published_at": <date>
    "pledge_url": <string>
    "pledge_sum": <int>
    "patron_count": <int>
    "creation_count": <int>
    "outstanding_payment_amount_cents": <int>
  }
  "relationships": {
    "creator": ...<user>...
    "rewards": [ ...<reward>, <reward>, ... ]
    "goals": [ ...<goal>, <goal>, ... ]
    "pledges": [ ...<pledge>, <pledge>, ... ]
  }
}
```

## Pledge

```typescript
{
  "type": "pledge"
  "id": <string>
  "attributes": {
    "amount_cents": <int> // Amount cents in the currency used by the patron
    "created_at": <date>
    "currency": <string> // Currency code of the pledge event (USD, GBP, EUR etc.)
    "declined_since": <date>
    "patron_pays_fees": <bool>
    "pledge_cap_cents": <int>
    // optional properties
    "total_historical_amount_cents": <int>
    "is_paused": <bool>
    "status": <string> // The status of this pledge (valid, declined, pending, disabled)
    "has_shipping_address": <bool>
  }
  "relationships": {
    "patron": ...<user>...
    "reward": ...<reward>... // Tier associated with this pledge
    "creator": ...<user>...
    "address": ...<address>...
  }
}
```

Also, you will want to add a [webhook](https://docs.patreon.com/#webhooks) to notify your app when the user stops the membership, you can remove any privileges / roles

```typescript
{
  "data": {
    "attributes": {
      "amount_cents": 250,
      "created_at": "2015-05-18T23:50:42+00:00",
      "declined_since": null,
      "patron_pays_fees": false,
      "pledge_cap_cents": null
    },
    "id": "1",
    "relationships": {
      "address": {
        "data": null
      },
      "card": {
        "data": null
      },
      "creator": {
        "data": {
          "id": "3024102",
          "type": "user"
        },
        "links": {
          "related": "https://www.patreon.com/api/user/3024102"
        }
      },
      "patron": {
        "data": {
          "id": "32187",
          "type": "user"
        },
        "links": {
          "related": "https://www.patreon.com/api/user/32187"
        }
      },
      "reward": {
        "data": {
          "id": "599336",
          "type": "reward"
        },
        "links": {
          "related": "https://www.patreon.com/api/rewards/599336"
        }
      }
    },
    "type": "pledge"
  },
  "included": [{ ** * Creator Object ** *
    },
    { ** * Patron Object ** *
    },
    { ** * Reward Object ** *
    },
  ]
}
```

**Comment by OKNoah** [comment]
I don't think I've seen a situation quite like this:

> Dear Patreon Developer Community —
>
> Over the past few months, we’ve heard creators and developers express that we haven’t been supporting our API as best we can. We want to be transparent and honest about our plans going forward. The news right now is that due to **resource constraints**, **we can no longer actively maintain the Developer API with Patreon engineering resources**. All endpoints will function as normal, however this means there won’t be new versions, ongoing maintenance, or updates. **Any resources dedicated to the API will be for fixing critical issues affecting uptime.**
>
> This doesn’t mean the API won’t work and developers tap into unmaintained API’s quite frequently, but **we want to make sure you make informed decisions based on our transparent plans for the future**. The last thing we want is for folks expecting to get assistance from us and being disappointed.
>
> Over the coming weeks, we’ll be migrating our developer forum 20 over to our main Patreon Community Forum 12. We’ll be able to keep the threads in developer forum and this is our official community forum with a lot more participation from both creators and the Patreon team. On the main forum, you’ll continue to have access to peer to peer support from other developers as well as resources available to help you run your membership. We’ll keep you updated on the timing and details of this migration.
>
> **We hope to be able to return to active support and continued updates** to the API in the future, but for now it will remain intact but in a unmaintained state.
>
> Patreon Team
[Emphasis added]

All I know of Patreon's operations is that it's raised over $250 million in funding and had $55 million revenue last year, a figure that doubles each year. Saying they lack the resources to maintain their endpoints just confuses me. I'd think a pseudo-payment processor would architect their service as a service, and develop and use the same APIs and libraries they release publicly internally as well. Just weird the API can be abandoned at all, as if it's a completely separate thing from the rest of their systems?

Then of course the ambiguity about what "abandoning" an API means, while they still imply they'll "fix critical issues affecting uptime".

> **we want to make sure you make informed decisions based on our transparent plans for the future**.

No one can make an informed decision with only these details. Will the API be deprecated and shut down? If so, when exactly. If they get what they hope for when they say "We hope to be able to return to active support and continued updates", when could that realistically happen? I would be much more confident in this service if they created a tentative timeline at least.

I'll ask again here what I said elsewhere:

> anyone got reccomendations for payment processors? i wanted to use Patreon because it seems to have excellent good will amongst people. almost has more of a altruistic energy to it, you "support" people and projects rather than simply buy a product from a corporation with your Visa. any other options out there like this? (i ask because i'm not confident in Patreon supporting their APIs much longer)
